R/exposure_metrics.R

```diff
@@ -132,3 +132,71 @@ calc_auc_from_regimen <- function(
     target_design$type
   )
 }
+
+#' Calculate time to target attainment
+#'
+#' @param regimen PKPDsim regimen object
+#' @param target_design target design, created using `create_target_design()`
+#' @param auc_comp compartment to look for AUC
+#' @param model PKPDsim model object
+#' @param terminal_interval duration of final dosing interval
+#' @param ... arguments passed on to `PKPDsim::sim`
+#' @return a numeric value indicating the time, in hours, if a dosing interval
+#'   contained the target metric within range. Otherwise, returns NA.
+#' @export
+calc_time_to_target <- function(
+    regimen,
+    target_design,
+    auc_comp,
+    model,
+    terminal_interval = 24,
+    ...
+) {
+  target_type <- match.arg(tolower(target_design$type), mipd_target_types())
+
+  iov <- PKPDsim::get_model_iov(model)
+  if (is.null(iov[["bins"]])) iov[["bins"]] <- c(0, 9999)
+
+  t_obs_target <- c(regimen$dose_times, tail(regimen$dose_times, 1) + terminal_interval)
+  sim_target <- PKPDsim::sim(
+    model,
+    regimen = regimen,
+    only_obs = FALSE,
+    t_obs = t_obs_target,
+    ...
+  )
+
+  # Which dose resulted in an exposure metric on-target?
+  dose_idx <- NULL
+  if (grepl("auc\\d+", target_type)){ # auc24 or auc12
+    auc_num <- as.numeric(gsub(".*auc(\\d+).*", "\\1", target_type, ignore.case = FALSE))
+    aucs_target <- sim_target$y[sim_target$comp == auc_comp]
+    auc_target_X <- auc_num * (diff(aucs_target) / diff(t_obs_target))
+    aucs_on_target <- which(
+      auc_target_X >= target_design$min & auc_target_X <= target_design$max
+    )
+    if (length(aucs_on_target) > 0) {
+      # return time of *start* of dosing interval that gives target
+      # plus infusion length
+      dose_idx <- min(aucs_on_target)
+    }
+  } else if (target_type %in% c("trough", "cmin")){
+    cmin_target <- sim_target$y[sim_target$comp == "obs"]
+    cmins_on_target <- which(
+      cmin_target >= target_design$min & cmin_target <= target_design$max
+    )
+    if (length(cmins_on_target) > 0) {
+      # return time of dose prior to trough in target (plus inf leng)
+      # i.e., the obs time before the first trough on target, or the
+      # first dose.
+      dose_idx <- pmax(min(cmins_on_target)-1,1)
+    }
+  }
+
+  # Return end of infusion of identified dose, or NA if no dose found
+  if (!is.null(dose_idx)) {
+    regimen$dose_times[dose_idx] + regimen$t_inf[dose_idx]
+  } else {
+    NA_real_
```

Comment:
Not because I think it is _wrong_ but more out of curiosity: why did you settle on NA_real_ as a return value if the patient never achieves target, versus e.g., Inf?

Reply:
Good point. I like this flow:

if not enough data, return NA_real_
if dose identified, return end of infusion
if no dose found, return Inf